src/main.cpp

```diff
@@ -28,20 +27,47 @@ int Main(vector<string> args)
 		%whitespace <- [ \t]*
 	)";
 
+	// create parser
 	parser parser(syntax);
 
 	if (!parser){
 		cout << "error creating grammar" << endl;
 		return EXIT_FAILURE;
 	}
 
-	parser.enable_packrat_parsing(); // Enable packrat parsing.
+	// create callbacks for debugging
+	auto rules = parser.get_rule_names();
 
-	int val;
+	for(auto r:rules){
+		parser[r.c_str()].enter = [r](const char* s, size_t n, any& dt) {
+			auto& indent = *dt.get<int*>();
+			for(int i=0; i<indent; i++){cout << "|  ";}
+			string beginning(s, n);
+			if(beginning.length()>10){beginning = beginning.substr(0,10) + "...";}
+			cout << r << " => \"" << beginning << "\"?" << endl;
+			indent++;
+		};
+
+		parser[r.c_str()].leave = [r](const char* s, size_t n, bool match, any& dt) {
+			auto& indent = *dt.get<int*>();
+			indent--;
+			for(int i=0; i<indent; i++){cout << "|  ";}
+			if(match){cout << "`-> match" << endl;}
+			else{cout << "`-> failed" << endl;}
+		};
+	}
 	parser.log = [&](size_t ln, size_t col, const string& msg) {
 		cout << "(" << ln << "," << col << ") " << msg;
 	};
-	bool success = parser.parse(" 10 * 5 - 1 ", val);
+
+	// parse string
+	parser.enable_packrat_parsing();
+
+	int val;
+
+	int indent = 0;
+	any dt = &indent;
```

Comment:
From a quick look, it appears that `dt` is always passed by reference inside peglib. So using a pointer is not required, it should suffice to use

    any dt = 0;